Photon Monolithic build fails on 0.7.0-rc.3

## Bug Report
When building locally against release/0.7.0-rc.3 a non-modular/monolithic build fails with an error.

`make clean all PLATFORM=photon MODULAR=n`

### Expected Behavior

A single binary containing the full system should be built.

### Observed Behavior

A compiler error is reported.

> ../build/target/hal/platform-6/\libhal.a(heap_4_lock.o):(.data.ucHeapEnd+0x0): undefined reference to `link_heap_location_end'
> ../build/target/hal/platform-6/\libhal.a(heap_4_lock.o):(.data.ucHeap+0x0): undefined reference to `link_heap_location'
> collect2.exe: error: ld returned 1 exit status
> make: *** [../build/target/main/platform-6/main.elf] Error 1

### Steps to Reproduce

* arm-none-eabi-gcc (GNU Tools for ARM Embedded Processors) 5.3.1 20160307 (release) [ARM/embedded-5-branch revision 234589]
* Windows 10

* `git checkout release/v0.7.0-rc.3`
* `cd main`
* `make clean all PLATFORM=photon MODULAR=n`

Hi!

I'm just trying out the latest RC with this fix:

```shell
git log -10 --pretty=oneline spark/release/v0.7.0-rc.3..
c9b671701a1a36c47ed026f1ff92d5b239141277 link_heap_location and link_heap_location_end definitions missing
```
then trying out following command `make clean all PLATFORM=photon MODULAR=n` which ends with this error:

```
Building c file: src/photon/core_hal.c
Invoking: ARM GCC C Compiler
mkdir -p ../build/target/hal/platform-6/./src/photon/
arm-none-eabi-gcc -DSTM32_DEVICE -DSTM32F2XX -DPLATFORM_THREADING=1 -DPLATFORM_ID=6 -DPLATFORM_NAME=photon -DUSBD_VID_SPARK=0x2B04 -DUSBD_PID_DFU=0xD006 -DUSBD_PID_CDC=0xC006 -g3 -gdwarf-2 -Os -mcpu=cortex-m3 -mthumb -DUSE_STDPERIPH_DRIVER -DDFU_BUILD_ENABLE -fno-builtin-memcpy -fno-builtin-memcmp -fno-builtin-memset -DMBEDTLS_CONFIG_FILE="<mbedtls_config.h>" -DPRODUCT_ID=6 -DPRODUCT_FIRMWARE_VERSION=65535 -DSYSTEM_VERSION_STRING=0.7.0-rc.3 -DRELEASE_BUILD -Werror -I./inc -I./shared -I./src/photon -I./src/stm32f2xx -I./src/stm32 -I./src/photon/api -I./src/photon/include -I./src/photon/wiced/security/BESL/host/WICED/ -I./src/photon/wiced/security/BESL/include -I./src/photon/wiced/security/BESL -I./src/photon/wiced/security/BESL/crypto -I./src/photon/wiced/WWD/include/ -I./src/photon/wiced/platform/include/ -I./src/photon/wiced/platform/GCC/ -I./src/photon/wiced/security/BESL/supplicant/ -I./src/photon/libraries/crypto -I../platform/shared/inc -I../platform/MCU/STM32F2xx/CMSIS/Include -I../platform/MCU/STM32F2xx/CMSIS/Device/ST/Include -I../platform/MCU/STM32F2xx/SPARK_Firmware_Driver/inc -I../platform/MCU/shared/STM32/inc -I../platform/MCU/STM32F2xx/STM32_StdPeriph_Driver/inc -I../platform/MCU/STM32F2xx/STM32_USB_Device_Driver/inc -I../platform/MCU/STM32F2xx/STM32_USB_Host_Driver/inc -I../platform/MCU/STM32F2xx/STM32_USB_OTG_Driver/inc -I../services/inc -I../dynalib/inc -I../bootloader/src/photon -I../bootloader/../hal/src/photon/wiced/platform/MCU -I../bootloader/../hal/src/photon/wiced/platform/MCU/STM32F2xx/WAF -I../bootloader/src/stm32f2xx -I../wiring/inc -I../crypto/inc -I../crypto/mbedtls/include -I../communication/src -I./src/photon/include -I./src/photon/libraries/daemons/DNS_redirect -I./src/photon/libraries/daemons/HTTP_server -I./src/photon/libraries/protocols/DNS -I./src/photon/libraries/utilities/linked_list -I./src/photon/libraries/utilities/ring_buffer -I./src/photon/platforms/BCM9WCDUSI09 -I./src/photon/wiced -I./src/photon/wiced/RTOS/FreeRTOS -I./src/photon/wiced/RTOS/FreeRTOS/WICED -I./src/photon/wiced/RTOS/FreeRTOS/WWD -I./src/photon/wiced/RTOS/FreeRTOS/WWD/ARM_CM3 -I./src/photon/wiced/RTOS/FreeRTOS/ver8.2.1/Source/include -I./src/photon/wiced/RTOS/FreeRTOS/ver8.2.1/Source/portable/GCC/ARM_CM3 -I./src/photon/wiced/WWD -I./src/photon/wiced/internal -I./src/photon/wiced/network/LwIP -I./src/photon/wiced/network/LwIP/WICED -I./src/photon/wiced/network/LwIP/WWD -I./src/photon/wiced/network/LwIP/WWD/FreeRTOS -I./src/photon/wiced/network/LwIP/ver1.4.1/src/include -I./src/photon/wiced/network/LwIP/ver1.4.1/src/include/ipv4 -I./src/photon/wiced/platform/ARM_CM3 -I./src/photon/wiced/platform/ARM_CM3/CMSIS -I./src/photon/wiced/platform/GCC -I./src/photon/wiced/platform/MCU -I./src/photon/wiced/platform/MCU/STM32F2xx -I./src/photon/wiced/platform/MCU/STM32F2xx/WAF -I./src/photon/wiced/platform/MCU/STM32F2xx/peripherals -I./src/photon/wiced/platform/include -I./src/photon/wiced/security/BESL/crypto/ -I./src/photon/wiced/security/BESL/crypto/ -I./src/photon/wiced/security/BESL/crypto_open/ -I./src/photon/wiced/security/BESL/crypto_open/ -I./src/photon/wiced/security/BESL/crypto_open/ -I./src/photon/wiced/security/BESL/crypto_open/ -I./src/photon/wiced/security/BESL/crypto_open/ -I./src/photon/wiced/security/BESL/crypto_open/ -I./src/photon/wiced/security/BESL/crypto_open/ -I./src/photon/wiced/security/BESL/crypto_open/ -I./src/photon/wiced/security/BESL/crypto_open/ -I./src/photon/wiced/security/BESL/crypto_open/ -I./src/photon/wiced/security/BESL/crypto_open/ -I./src/photon/wiced/security/BESL/crypto_open/ -I./src/photon/wiced/security/BESL/crypto_open/ -I./src/photon/wiced/security/BESL/crypto_open/ -I./src/photon/wiced/security/BESL/crypto_open/ -I./src/photon/wiced/security/BESL/crypto_open/ -I./src/photon/wiced/security/BESL/crypto_open/ -I./src/photon/wiced/security/BESL/crypto_open/ -I./src/photon/wiced/security/BESL/crypto_open/ -I./src/photon/wiced/security/BESL/crypto_open/ -I./src/photon/wiced/security/BESL/crypto_open/ -I./src/photon/wiced/security/BESL/crypto_open/ -I./src/photon/wiced/security/BESL/crypto_open/ed25519/ -I./src/photon/wiced/security/BESL/crypto_open/ed25519/ -I./src/photon/wiced/security/BESL/crypto_open/ed25519/ -I./src/photon/wiced/security/BESL/crypto_open/ed25519/ -I./src/photon/wiced/security/BESL/crypto_open/ed25519/ -I./src/photon/wiced/security/BESL/crypto_open/ed25519/ -I./src/photon/wiced/security/BESL/crypto_open/ed25519/ -I./src/photon/wiced/security/BESL/crypto_open/ed25519/ -I./src/photon/wiced/security/BESL/crypto_open/ed25519/ -I./src/photon/wiced/security/BESL/crypto_open/ed25519/ -I./src/photon/wiced/security/BESL/crypto_open/ed25519/ -I./src/photon/wiced/security/BESL/crypto_open/ed25519/ -I./src/photon/wiced/security/BESL/crypto_open/ed25519/ -I./src/photon/wiced/security/BESL/crypto_open/ed25519/ -I./src/photon/wiced/security/BESL/crypto_open/ed25519/ -I./src/photon/wiced/security/BESL/crypto_open/unit/ -I./src/photon/wiced/security/BESL/DTLS/ -I./src/photon/wiced/security/BESL/DTLS/ -I./src/photon/wiced/security/BESL/host/WICED/ -I./src/photon/wiced/security/BESL/host/WICED/ -I./src/photon/wiced/security/BESL/host/WICED/ -I./src/photon/wiced/security/BESL/host/WICED/ -I./src/photon/wiced/security/BESL/host/WICED/ -I./src/photon/wiced/security/BESL/host/WICED/ -I./src/photon/wiced/security/BESL/host/WICED/ -I./src/photon/wiced/security/BESL/host/WICED/ -I./src/photon/wiced/security/BESL/host/WICED/ -I./src/photon/wiced/security/BESL/host/WICED/ -I./src/photon/wiced/security/BESL/host/WICED/ -I./src/photon/wiced/security/BESL/host/WICED/ -I./src/photon/wiced/security/BESL/host/WICED/ -I./src/photon/wiced/security/BESL/include/ -I./src/photon/wiced/security/BESL/include/ -I./src/photon/wiced/security/BESL/include/ -I./src/photon/wiced/security/BESL/include/ -I./src/photon/wiced/security/BESL/include/ -I./src/photon/wiced/security/BESL/include/ -I./src/photon/wiced/security/BESL/include/ -I./src/photon/wiced/security/BESL/P2P/ -I./src/photon/wiced/security/BESL/P2P/ -I./src/photon/wiced/security/BESL/P2P/ -I./src/photon/wiced/security/BESL/supplicant/ -I./src/photon/wiced/security/BESL/supplicant/ -I./src/photon/wiced/security/BESL/TLS/ -I./src/photon/wiced/security/BESL/TLS/ -I./src/photon/wiced/security/BESL/WPS/ -I./src/photon/wiced/security/BESL/WPS/ -I./src/photon/wiced/security/BESL/WPS/ -I./src/photon/wiced/security/BESL/WPS/ -I./src/photon/wiced/WWD/include/ -I./src/photon/wiced/WWD/include/ -I./src/photon/wiced/WWD/include/ -I./src/photon/wiced/WWD/include/ -I./src/photon/wiced/WWD/include/ -I./src/photon/wiced/WWD/include/ -I./src/photon/wiced/WWD/include/ -I./src/photon/wiced/WWD/include/ -I./src/photon/wiced/WWD/include/ -I./src/photon/wiced/WWD/include/ -I./src/photon/wiced/WWD/include/network/ -I./src/photon/wiced/WWD/include/network/ -I./src/photon/wiced/WWD/include/network/ -I./src/photon/wiced/WWD/include/platform/ -I./src/photon/wiced/WWD/include/platform/ -I./src/photon/wiced/WWD/include/platform/ -I./src/photon/wiced/WWD/include/platform/ -I./src/photon/wiced/WWD/include/platform/ -I./src/photon/wiced/WWD/include/RTOS/ -I./src/photon/wiced/WWD/internal/ -I./src/photon/wiced/WWD/internal/ -I./src/photon/wiced/WWD/internal/ -I./src/photon/wiced/WWD/internal/ -I./src/photon/wiced/WWD/internal/ -I./src/photon/wiced/WWD/internal/ -I./src/photon/wiced/WWD/internal/ -I./src/photon/wiced/WWD/internal/ -I./src/photon/wiced/WWD/internal/bus_protocols/ -I./src/photon/wiced/WWD/internal/bus_protocols/SDIO/ -I./src/photon/wiced/WWD/internal/bus_protocols/SoC/4390/ -I./src/photon/wiced/WWD/internal/bus_protocols/SoC/43909/ -I./src/photon/wiced/WWD/internal/bus_protocols/SPI/ -I./src/photon/wiced/WWD/internal/chips/4334x/ -I./src/photon/wiced/WWD/internal/chips/43362/ -I./src/photon/wiced/WWD/internal/chips/4343x/ -I./src/photon/wiced/WWD/internal/chips/4390/ -I./src/photon/wiced/WWD/internal/chips/4390x/ -I./src/stm32f2xx -I./src/stm32 -I. -MD -MP -MF ../build/target/hal/platform-6/./src/photon/core_hal.o.d -ffunction-sections -fdata-sections -Wall -Wno-switch -Wno-error=deprecated-declarations -fmessage-length=0 -fno-strict-aliasing -DSPARK=1 -DPARTICLE=1 -Wundef -DSTART_DFU_FLASHER_SERIAL_SPEED=14400 -DSTART_YMODEM_FLASHER_SERIAL_SPEED=28800 -DBOOTLOADER_SDK_3_3_0_PARTICLE -DPARTICLE_DCT_COMPATIBILITY -DSFLASH_APPS_HEADER_LOC=0x0000 -DUSE_STDPERIPH_DRIVER -D_STM32F215RGT6_ -D_STM3x_ -D_STM32x_ -DMAX_WATCHDOG_TIMEOUT_SECONDS=22 -DFIRMWARE_WITH_PMK_CALC_SUPPORT -DADD_LWIP_EAPOL_SUPPORT -DNXD_EXTENDED_BSD_SOCKET_SUPPORT -DOPENSSL -DSTDC_HEADERS -DUSE_SRP_SHA_512 -DADD_NETX_EAPOL_SUPPORT -DUSE_MICRORNG -DWWD_STARTUP_DELAY=10 -DBOOTLOADER_MAGIC_NUMBER=0x4d435242 -DNETWORK_NetX=1 -DNetX_VERSION=\"v5.5_sp1\" -DNX_INCLUDE_USER_DEFINE_FILE -D__fd_set_defined -DSYS_TIME_H_AVAILABLE -DRTOS_ThreadX=1 -DThreadX_VERSION=\"v5.6\" -DTX_INCLUDE_USER_DEFINE_FILE -DWWD_DIRECT_RESOURCES -D_WINSOCK_H -D_POSIX_C_SOURCE=200809 -DLOG_MODULE_CATEGORY="\"hal\""  -Wno-pointer-sign -std=gnu99 -c -o ../build/target/hal/platform-6/./src/photon/core_hal.o src/photon/core_hal.c
In file included from ./src/stm32f2xx/core_hal_stm32f2xx.h:4:0,
                 from src/photon/core_hal.c:27:
../platform/MCU/STM32F2xx/SPARK_Firmware_Driver/inc/hw_config.h:87:5: error: "MODULE_FUNCTION" is not defined [-Werror=undef]
 #if MODULE_FUNCTION != MOD_FUNC_BOOTLOADER
     ^~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
```

```
arm-none-eabi-gcc (GNU Tools for ARM Embedded Processors 6-2017-q2-update) 6.3.1 20170620 (release) [ARM/embedded-6-branch revision 249437]
Copyright (C) 2016 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

Compiler version (un)related issue?

I get exactly the same error even with older GCC:

```
arm-none-eabi-gcc (GNU Tools for ARM Embedded Processors) 5.4.1 20160919 (release) [ARM/embedded-5-branch revision 240496]
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

```
In file included from ./src/stm32f2xx/core_hal_stm32f2xx.h:4:0,
                 from src/photon/core_hal.c:27:
../platform/MCU/STM32F2xx/SPARK_Firmware_Driver/inc/hw_config.h:87:5: error: "MODULE_FUNCTION" is not defined [-Werror=undef]
 #if MODULE_FUNCTION != MOD_FUNC_BOOTLOADER
     ^
cc1: all warnings being treated as errors
make[1]: *** [../build/target/hal/platform-6/./src/photon/core_hal.o] Error 1
```

--- a/hal/src/photon/wiced/platform/MCU/STM32F2xx/GCC/app_no_bootloader.ld
+++ b/hal/src/photon/wiced/platform/MCU/STM32F2xx/GCC/app_no_bootloader.ld
@@ -167,6 +167,9 @@ SECTIONS
     INCLUDE backup_ram_user.ld
     INCLUDE backup_ram_system.ld
 
+    link_heap_location = _heap;
+    link_heap_location_end = _eheap;
+
     /DISCARD/ :
     {
         *(.ARM.attributes*)

